### PR TITLE
Feat/async callbacks

### PR DESCRIPTION
**Pull request by o-bieliaiev**

### CODE REVIEW

File: .env.development [hunk 1]
```diff
@@ -7,3 +7,6 @@ VITE_AUTH_AUDIENCEVITE_AUTH_CLIENT_ID=some-id
 VITE_DID_API_URL=https://api-dev.d-id.com
 VITE_LAUNCHDARKLY_CLIENT_KEY=some-key
 VITE_MIXPANEL_KEY=some-key
+VITE_WS_ENDPOINT=wss://notifications-dev.d-id.com
+VITE_CLIENT_KEY=Z29vZ2xlLW9hdXRoMnwxMTc1MzA5OTE1NTk3MzcxNTc5MjM6RmJaY052cm1DUENCblBCWHl0UXNW
```
Comment (o-bieliaiev) [inline .env.development:11]: move to .env

File: src/lib/api/ratings.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { Auth, RatingEntity, RatingPayload } from '$/index';
+import { createClient } from './getClient';
+
+
+export function createRatingssApi(auth: Auth, host = 'https://api.d-id.com') {
```
Comment (o-bieliaiev) [inline src/lib/api/ratings.ts:5]: use CONST here

File: src/lib/connectToSocket.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    return randomID;
+}
+
+function connect(options: Options): Promise<WebSocket> {
+    return new Promise<WebSocket>((resolve, reject) => {
+        const { callbacks, host, auth } = options;
+        const { onMessage = null, onOpen = null, onClose = null, onError = null } = callbacks || {};
+        // TODO discuss with Sagi. If I create socket with randomID, haven't recieve messages
```
Comment (o-bieliaiev) [inline src/lib/connectToSocket.ts:42]: rename TODO

File: src/lib/connectToSocket.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        };
+
+        socket.onopen = e => {
+            onOpen?.(e);
+            resolve(socket);
+        };
+    });
+}
+
+export async function connectToSocket(options: Options): Promise<WebSocket> {
```
Comment (o-bieliaiev) [inline src/lib/connectToSocket.ts:63]: remove to export

Comment (o-bieliaiev) [reply src/lib/connectToSocket.ts:63]: rename to connectWithRetries

File: src/lib/connectToSocket.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    return socket
+}
+
+
+export async function SocketManager(auth: Auth, host: string = socketHost): Promise<SocketManager> {
+    let socket: WebSocket | null
+    let messageCallbacks: ((data: any) => void)[] = [];
+    let socketManager: SocketManager = {
+        terminate: () => socketManager.socket?.close(),
+        connect: async () => {
```
Comment (o-bieliaiev) [inline src/lib/connectToSocket.ts:88]: move connect logic out of method an remove

File: src/lib/connectToSocket.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+            socket = await connectToSocket({
+                auth,
+                host,
+                callbacks: { onMessage: handleMessage },
```
Comment (o-bieliaiev) [inline src/lib/connectToSocket.ts:92]: move function declaration inside

File: src/lib/createAgentManager.ts [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+export async function createAgentManager(agentId: string, options: AgentManagerOptions): Promise<AgentsAPI> {
+    const baseURL = options.baseURL ? options.baseURL : CONSTANTS.baseURL;
+    const abortController: AbortController = new AbortController();
+    const agentsApi = createAgentsApi(options.auth, baseURL);
+
+    const agent = await agentsApi.getById(agentId);
+    const chat = await agentsApi.newChat(agentId);
+
+    const ratingsAPI = await createRatingssApi(options.auth, baseURL);
```
Comment (o-bieliaiev) [inline src/lib/createAgentManager.ts:54]: remove await

Comment (o-bieliaiev) [reply src/lib/createAgentManager.ts:54]: ratignsAPI after agentsApi

File: src/lib/createAgentManager.ts [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+ * * @returns {Promise<AgentsAPI>} - A promise that resolves to an instance of the AgentsAPI interface.
+ *
+ * @throws {Error} Throws an error if the agent is not initialized.
+ *
+ * @example
+ * const agentManager = await createAgentManager('id-agent123', { auth: { type: 'key', clientKey: '123', externalId: '123' } });
+ */
+
+export async function createAgentManager(agentId: string, options: AgentManagerOptions): Promise<AgentsAPI> {
+    const baseURL = options.baseURL ? options.baseURL : CONSTANTS.baseURL;
```
Comment (HaMMeRSI) [inline src/lib/createAgentManager.ts:47]: const baseURL = options.baseURL || CONSTANTS.baseURL;

File: src/lib/connectToSocket.ts [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+            });
+
+            socketManager.socket = socket;  
+
+            return socket;
+        },
+        subscribeToEvents: (callback: (data: any) => void) => {
```
Comment (HaMMeRSI) [inline src/lib/connectToSocket.ts:99]: if (!socketManager.socket.isconenct) {
                 throw new Error('Socket is not connected. Please call connect() first.');

            } 

                messageCallbacks.push(callback);

Comment (HaMMeRSI) [reply src/lib/connectToSocket.ts:99]: Make generic

File: src/lib/createAgentManager.ts [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        },
+        onVideoEvents(callback: VideoStateChangeCallback) {
+            streamingAPI.addCallback('onVideoStateChange', callback);
+        },
+    };
+
+    return resultAPI;
+}
+
+function filterCallbacks(options: AgentManagerOptions) {
```
Comment (HaMMeRSI) [inline src/lib/createAgentManager.ts:138]: not necessary

File: src/lib/createStreamingManager.ts [hunk 11]
```diff
@@ -16,6 +25,7 @@ export async function createStreamingManager<T extends CreateStreamOptions>(
 ) {
     _debug = debug;
     let srcObject: MediaStream | null = null;
+    const callbacksObj: any = callbacks ? callbacks : {};
```
[the remark this answers is not in the document]
Comment (HaMMeRSI) [reply src/lib/createStreamingManager.ts:28]: const callbacksObj: any = {...callbacks}

**Review by HaMMeRSI — CHANGES_REQUESTED** [review]

File: src/lib/connectToSocket.ts (hunk 9 above)
Comment (o-bieliaiev) [reply src/lib/connectToSocket.ts]: made generic, removed case when socket could not exist

File: src/lib/createStreamingManager.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (o-bieliaiev) [reply src/lib/createStreamingManager.ts]: done

File: src/lib/createAgentManager.ts (hunk 10 above)
Comment (o-bieliaiev) [reply src/lib/createAgentManager.ts]: removed

File: src/lib/createStreamingManager.ts [hunk 12]
```diff
@@ -90,13 +116,33 @@ export async function createStreamingManager<T extends CreateStreamOptions>(
...
+         * @param eventName 
+         * @param callback 
+         */
+        onCallback<T extends ManagerCallbackKeys>(eventName: T, callback: ManagerCallbacks[T]): void {
+            callbacksObj[eventName] = callback;
+        },
+        /**
+         * existing callback is internal method to pass calbacks added after create in new connection
+         */
+        getCallbacks() {
```
Comment (HaMMeRSI) [inline src/lib/createStreamingManager.ts:143]: why do we need this?

**Review by HaMMeRSI — CHANGES_REQUESTED** [review]

File: src/lib/createAgentManager.ts [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+    return {
+        agent,
+        async reconnectToChat() {
+            const { chat: newChat, streamingManager: newStreamingManager } = await initializeStreamAndChat(
+                agent,
+                options,
+                agentsApi,
+                chat
+            );
+            chat = newChat;
```
[the remark this answers is not in the document]
Comment (o-bieliaiev) [reply src/lib/createAgentManager.ts]: removed

File: src/lib/createStreamingManager.ts (hunk 12 above)
Comment (o-bieliaiev) [reply src/lib/createStreamingManager.ts]: nope, removed

File: src/lib/connectToSocket.ts [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+export async function SocketManager(auth: Auth, host: string = didSocketApiUrl): Promise<SocketManager> {
+    const messageCallbacks: ((data: any) => void)[] = [];
+    const socket: WebSocket = await connectWithRetries({
+        auth,
+        host,
+        callbacks: {
+            onMessage: (event: MessageEvent) => {
+                console.log('event', event);
```
[the remark this answers is not in the document]
Comment (o-bieliaiev) [reply src/lib/connectToSocket.ts]: done

File: src/lib/connectToSocket.ts [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+                messageCallbacks.forEach(callback => callback(event));
+            },
+        },
+    });
+
+    return {
+        socket,
+        terminate: () => socket.close(),
+        subscribeToEvents: <T>(callback: (data: T) => void) => {
```
[the remark this answers is not in the document]
Comment (o-bieliaiev) [reply src/lib/connectToSocket.ts:95]: done

File: src/lib/createAgentManager.ts [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+        },
+        terminate() {
+            abortController.abort();
+            socketManager.terminate();
+            return streamingManager.terminate();
+        },
+        chatId: chat.id,
+        chat(messages: Message[]) {
+            // should be diffrent after reconnect
+            console.log(streamingManager.sessionId)
```
[the remark this answers is not in the document]
Comment (o-bieliaiev) [reply src/lib/createAgentManager.ts]: removed

File: src/lib/createAgentManager.ts [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+
+    const agent = await agentsApi.getById(agentId);
+    const socketManager = await SocketManager(options.auth);
+    let { chat, streamingManager } = await initializeStreamAndChat(agent, options, agentsApi);
+
+    return {
+        agent,
+        async reconnectToChat() {
+            await initializeStreamAndChat(agent, options, agentsApi, chat?.id).then(result => {
+                chat = result.chat;
```
[the remark this answers is not in the document]
Comment (o-bieliaiev) [reply src/lib/createAgentManager.ts]: agree, done

**Review by HaMMeRSI — APPROVED** [review]

**Pull request merged**